src/coreclr/jit/importer.cpp

```diff
@@ -18912,15 +18912,15 @@ void Compiler::impMakeDiscretionaryInlineObservations(InlineInfo* pInlineInfo, I
         }
     }
 
-    // Check if callee is a generic method when its caller is not
-    bool calleeIsGeneric = (info.compMethodInfo->args.callConv & CORINFO_CALLCONV_GENERIC) ||
-                           (info.compClassAttr & CORINFO_FLG_SHAREDINST);
-    bool callerIsGeneric = (rootCompiler->info.compMethodInfo->args.callConv & CORINFO_CALLCONV_GENERIC) ||
-                           (rootCompiler->info.compClassAttr & CORINFO_FLG_SHAREDINST);
-
-    if (!callerIsGeneric && calleeIsGeneric)
+    if ((info.compMethodInfo->args.callConv & CORINFO_CALLCONV_GENERIC) ||
+        (info.compClassAttr & CORINFO_FLG_SHAREDINST))
+    {
+        inlineResult->Note(InlineObservation::CALLEE_GENERIC);
+    }
+    else if ((rootCompiler->info.compMethodInfo->args.callConv & CORINFO_CALLCONV_GENERIC) ||
+             (rootCompiler->info.compClassAttr & CORINFO_FLG_SHAREDINST))
     {
-        inlineResult->Note(InlineObservation::CALLSITE_GENERIC_FROM_NONGENERIC);
+        inlineResult->Note(InlineObservation::CALLSITE_GENERIC);
```

Comment:
This is a caller attribute, isn't it? Also perhaps be explicit that the shared aspect is what makes it interesting.
```suggestion
        inlineResult->Note(InlineObservation::CALLER_SHARED_GENERIC);
```
Previously you were noting when a non-shared caller was invoking a shared callee; this still seems to be the interesting combination.  Shared callers invoking shared callees will be blocked in many cases, because dictionary lookups always refer to the root method's dictionary.

So is the plan to still detect this non-shared caller/ shared callee by the absence of an observation? I think I'd prefer we make an observation then deduce something because we didn't make an observation.

Reply:
Changed back to be NonGenericCallsGeneric. And it's not only about shared ones since it's also just checks method signature for <>.